pkg/scalers/aws_dynamodb_streams_scaler.go

```diff
@@ -0,0 +1,228 @@
+	metricType, err := GetMetricTargetType(config)
+	if err != nil {
+		return nil, fmt.Errorf("error getting scaler metric type: %s", err)
+	}
+
+	meta, err := parseAwsDynamoDBStreamsMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing dynamodb stream metadata: %s", err)
+	}
+
+	dbClient, dbStreamClient := createClientsForDynamoDBStreamsScaler(meta)
+
+	streamArn, err := getDynamoDBStreamsArn(ctx, dbClient, &meta.tableName)
+	if err != nil {
+		return nil, fmt.Errorf("error dynamodb stream arn: %s", err)
+	}
+
+	return &awsDynamoDBStreamsScaler{
+		metricType:     metricType,
+		metadata:       meta,
+		streamArn:      streamArn,
+		dbStreamClient: dbStreamClient,
+	}, nil
+}
+
+func parseAwsDynamoDBStreamsMetadata(config *ScalerConfig) (*awsDynamoDBStreamsMetadata, error) {
+	meta := awsDynamoDBStreamsMetadata{}
+	meta.targetShardCount = defaultTargetDBStreamsShardCount
+
+	if val, ok := config.TriggerMetadata["awsRegion"]; ok && val != "" {
+		meta.awsRegion = val
+	} else {
+		return nil, fmt.Errorf("no awsRegion given")
+	}
+
+	if val, ok := config.TriggerMetadata["tableName"]; ok && val != "" {
+		meta.tableName = val
+	} else {
+		return nil, fmt.Errorf("no tableName given")
+	}
+
+	if val, ok := config.TriggerMetadata["shardCount"]; ok && val != "" {
+		shardCount, err := strconv.ParseInt(val, 10, 64)
+		if err != nil {
+			meta.targetShardCount = defaultTargetDBStreamsShardCount
+			dynamodbStreamLog.Error(err, "error parsing dyanmodb stream metadata shardCount, using default %n", defaultTargetDBStreamsShardCount)
+		} else {
+			meta.targetShardCount = shardCount
+		}
+	}
+
+	auth, err := getAwsAuthorization(config.AuthParams, config.TriggerMetadata, config.ResolvedEnv)
+	if err != nil {
+		return nil, err
+	}
+
+	meta.awsAuthorization = auth
+	meta.scalerIndex = config.ScalerIndex
+
+	return &meta, nil
+}
+
+func createClientsForDynamoDBStreamsScaler(metadata *awsDynamoDBStreamsMetadata) (*dynamodb.DynamoDB, *dynamodbstreams.DynamoDBStreams) {
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String(metadata.awsRegion),
+	}))
+
+	var dbClient *dynamodb.DynamoDB
+	var dbStreamClient *dynamodbstreams.DynamoDBStreams
+
+	if metadata.awsAuthorization.podIdentityOwner {
+		creds := credentials.NewStaticCredentials(metadata.awsAuthorization.awsAccessKeyID, metadata.awsAuthorization.awsSecretAccessKey, metadata.awsAuthorization.awsSessionToken)
+		if metadata.awsAuthorization.awsRoleArn != "" {
+			creds = stscreds.NewCredentials(sess, metadata.awsAuthorization.awsRoleArn)
+		}
+		dbClient = dynamodb.New(sess, &aws.Config{
+			Region:      aws.String(metadata.awsRegion),
+			Credentials: creds,
+		})
+		dbStreamClient = dynamodbstreams.New(sess, &aws.Config{
+			Region:      aws.String(metadata.awsRegion),
+			Credentials: creds,
+		})
+	} else {
+		dbClient = dynamodb.New(sess, &aws.Config{
+			Region: aws.String(metadata.awsRegion),
+		})
+		dbStreamClient = dynamodbstreams.New(sess, &aws.Config{
```

Comment:
Just double checking, these clients doesn't have to be closed?

Reply:
No, it doesn't have to be closed

Reply:
Ack :)